### PR TITLE
Breaking down libraryId into name/version and validating the library …

### PR DESCRIPTION
**Pull request by raquelsa**
…against the catalog.

### CODE REVIEW

File: src/LibraryManager/Helpers/Extensions.cs [hunk 1]
```diff
@@ -22,42 +22,92 @@ public static class Extensions
         /// Returns true if the <see cref="ILibraryInstallationState"/> is valid.
         /// </summary>
         /// <param name="state">The state to test.</param>
-        /// <param name="errors">The errors contained in the <see cref="ILibraryInstallationState"/> if any.</param>
+        /// <param name="dependencies"></param>
         /// <returns>
         ///   <c>true</c> if the specified state is valid; otherwise, <c>false</c>.
```
Comment (raquelsa) [inline src/LibraryManager/Helpers/Extensions.cs]: Fix return type. It now returns an ILibraryOperationResult

File: test/LibraryManager.Mocks/Provider.cs [hunk 2]
```diff
@@ -100,5 +100,17 @@ public string GetSuggestedDestination(ILibrary library)
         {
             return library?.Name;
         }
+
```
Comment (raquelsa) [inline test/LibraryManager.Mocks/Provider.cs]: Remove this

File: test/LibraryManager.Test/CacheServiceTest.cs [hunk 3]
```diff
@@ -4,14 +4,13 @@
 using System;
 using System.Collections.Generic;
 using System.IO;
-using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
```
Comment (raquelsa) [inline test/LibraryManager.Test/CacheServiceTest.cs]: Remove this fiel changes

File: test/LibraryManager.Test/Providers/Cdnjs/CdnjsProviderTest.cs [hunk 4]
```diff
@@ -150,9 +150,7 @@ public void Cleanup()
 
             // Install library
             ILibraryOperationResult result = await _provider.InstallAsync(desiredState, CancellationToken.None).ConfigureAwait(false);
-            Assert.IsFalse(result.Success);
-            Assert.AreEqual(1, result.Errors.Count);
-            Assert.AreEqual("LIB007", result.Errors.First().Code);
+            Assert.IsTrue(result.Success);
```
Comment (raquelsa) [inline test/LibraryManager.Test/Providers/Cdnjs/CdnjsProviderTest.cs]: We calling InstallAsync from the provider itself, we know which provider to use.

File: src/LibraryManager/Providers/Cdnjs/CdnjsProvider.cs [hunk 5]
```diff
@@ -84,9 +85,10 @@ public ILibraryCatalog GetCatalog()
                 return LibraryOperationResult.FromCancelled(desiredState);
             }
 
-            if (!desiredState.IsValid(out IEnumerable<IError> errors))
+            ILibraryOperationResult validationResult = await desiredState.IsValidAsync(this);
```
Comment (prafullbhosale) [inline src/LibraryManager/Providers/Cdnjs/CdnjsProvider.cs]: Shouldn't this validation be done at the manifest level?

File: src/LibraryManager/Providers/Cdnjs/CdnjsProvider.cs [hunk 6]
```diff
@@ -237,35 +239,21 @@ public string GetSuggestedDestination(ILibrary library)
 
         private async Task<Stream> GetStreamAsync(ILibraryInstallationState state, string sourceFile, CancellationToken cancellationToken)
         {
-            string name = GetLibraryName(state.LibraryId);
-            string version = GetLibraryVersion(state.LibraryId);
+            (string name, string version) = LibraryNamingScheme.Instance.GetLibraryNameAndVersion(state.LibraryId);
```
Comment (prafullbhosale) [inline src/LibraryManager/Providers/Cdnjs/CdnjsProvider.cs]: You shouldn't have to do this. 
The `ILibraryInstallationState` has the name and version properties already.

File: src/LibraryManager/Providers/Cdnjs/CdnjsProvider.cs [hunk 7]
```diff
@@ -280,35 +268,39 @@ public string GetLibraryVersion(string libraryId)
...
-            string name = GetLibraryName(state.LibraryId);
-            string version = GetLibraryVersion(state.LibraryId);
-
-            string libraryDir = Path.Combine(CacheFolder, name);
 
             try
             {
-                List<CacheServiceMetadata> librariesMetadata = new List<CacheServiceMetadata>();
-                foreach (string sourceFile in state.Files)
+                (string name, string version) = LibraryNamingScheme.Instance.GetLibraryNameAndVersion(state.LibraryId);
```
Comment (prafullbhosale) [inline src/LibraryManager/Providers/Cdnjs/CdnjsProvider.cs]: Same as line 242.

File: src/LibraryManager/Providers/Cdnjs/CdnjsProvider.cs [hunk 8]
```diff
@@ -319,35 +311,35 @@ public string GetLibraryVersion(string libraryId)
...
-            string name = GetLibraryName(state.LibraryId);
-            string version = GetLibraryVersion(state.LibraryId);
-
-            string cacheDir = Path.Combine(CacheFolder, name, version);
-            string destinationDir = Path.Combine(HostInteraction.WorkingDirectory, state.DestinationPath);
-
             try
             {
-                foreach (string sourceFile in state.Files)
+                (string name, string version) = LibraryNamingScheme.Instance.GetLibraryNameAndVersion(state.LibraryId);
```
Comment (prafullbhosale) [inline src/LibraryManager/Providers/Cdnjs/CdnjsProvider.cs]: Same as line 242

File: src/LibraryManager/Providers/Unpkg/UnpkgProvider.cs [hunk 9]
```diff
@@ -109,27 +111,25 @@ public string GetSuggestedDestination(ILibrary library)
             }
 
             var tasks = new List<Task>();
-            string[] args = state.LibraryId.Split('@');
-            string name = args[0];
-            string version = args[1];
+            (string name, string version) = LibraryNamingScheme.Instance.GetLibraryNameAndVersion(state.LibraryId);
```
Comment (prafullbhosale) [inline src/LibraryManager/Providers/Unpkg/UnpkgProvider.cs]: Same as the other provider.

File: src/LibraryManager/Microsoft.Web.LibraryManager.csproj [hunk 10]
```diff
@@ -19,6 +19,9 @@
   <ItemGroup>
     <XliffResource Include="MultilingualResources\*.xlf" />
   </ItemGroup>
+  <ItemGroup>
```
Comment (prafullbhosale) [inline src/LibraryManager/Microsoft.Web.LibraryManager.csproj]: Can be removed.

File: test/LibraryManager.Test/LibraryInstallationStateTest.cs [hunk 11]
```diff
@@ -26,5 +52,155 @@ public void FromInterface()
...
+            {
+                ProviderId = "filesystem",
+                LibraryId = "http://glyphlist.azurewebsites.net/img/images/Flag.png",
+                DestinationPath = "lib",
+                Files = new[] { "foo.png" }
+            };
+
+            ILibraryOperationResult result = await state.IsValidAsync(_dependencies);
+
+            Assert.IsTrue(result.Success);
```
Comment (prafullbhosale) [inline test/LibraryManager.Test/LibraryInstallationStateTest.cs]: Could you add a comment to indicate why this is expected to succeed. 
Took me a while to recollect that FS provider supports renaming.

File: src/LibraryManager/Manifest.cs [hunk 12]
```diff
@@ -217,9 +217,10 @@ public Manifest Clone()
             };
 
             UpdateLibraryProviderAndDestination(desiredState, DefaultProvider, DefaultDestination);
-            if (!desiredState.IsValid(out IEnumerable<IError> errors))
+            ILibraryOperationResult validationResult = await desiredState.IsValidAsync(_dependencies);
+            if (!validationResult.Success)
             {
-                return new List<ILibraryOperationResult> { new LibraryOperationResult(desiredState, errors.ToArray()) };
+                return new List<ILibraryOperationResult> { validationResult };
```
Comment (jimmylewis) [inline src/LibraryManager/Manifest.cs]: >List<ILibraryOperationResult> [](start = 27, length = 29)

array

File: src/LibraryManager/Manifest.cs [hunk 13]
```diff
@@ -477,23 +478,27 @@ public void AddVersion(string version)
...
             if (libraries != null)
             {
-                foreach (ILibraryInstallationState state in libraries.Where(l => l.IsValid(out IEnumerable<IError> errors)))
+                foreach (ILibraryInstallationState state in libraries)
                 {
-                    IProvider provider = _dependencies.GetProvider(state.ProviderId);
-
-                    if (provider != null)
+                    ILibraryOperationResult validationResult = await state.IsValidAsync(_dependencies);
+                    if (validationResult.Success)
```
Comment (jimmylewis) [inline src/LibraryManager/Manifest.cs]: can you add an else with a Debug.Assert?

File: src/LibraryManager/Manifest.cs [hunk 14]
```diff
@@ -477,23 +478,27 @@ public void AddVersion(string version)
 
             if (libraries != null)
             {
-                foreach (ILibraryInstallationState state in libraries.Where(l => l.IsValid(out IEnumerable<IError> errors)))
+                foreach (ILibraryInstallationState state in libraries)
                 {
-                    IProvider provider = _dependencies.GetProvider(state.ProviderId);
-
-                    if (provider != null)
+                    ILibraryOperationResult validationResult = await state.IsValidAsync(_dependencies);
```
Comment (jimmylewis) [inline src/LibraryManager/Manifest.cs]: >await [](start = 63, length = 5)

should we batch these so that we can run them all at once and do a WhenAll?

Comment (jimmylewis) [inline src/LibraryManager/Manifest.cs]: >; [](start = 102, length = 1)

ConfigureAwait

File: src/LibraryManager/LibrariesValidator.cs [hunk 15]
```diff
@@ -46,7 +46,7 @@ public LibrariesValidator(IDependencies dependencies, string defaultDestination,
         {
             cancellationToken.ThrowIfCancellationRequested();
 
-            IEnumerable<ILibraryOperationResult> validateLibraries = ValidatePropertiesAsync(libraries, cancellationToken);
+            IEnumerable<ILibraryOperationResult> validateLibraries = await ValidatePropertiesAsync(libraries, cancellationToken);
```
Comment (jimmylewis) [inline src/LibraryManager/LibrariesValidator.cs]: ConfigureAwait

File: src/LibraryManager/LibrariesValidator.cs [hunk 16]
```diff
@@ -72,17 +72,18 @@ public LibrariesValidator(IDependencies dependencies, string defaultDestination,
...
+        public async Task<IEnumerable<ILibraryOperationResult>> ValidatePropertiesAsync(IEnumerable<ILibraryInstallationState> libraries, CancellationToken cancellationToken)
         {
             List<ILibraryOperationResult> validationStatus = new List<ILibraryOperationResult>();
 
             foreach (ILibraryInstallationState library in libraries)
             {
                 cancellationToken.ThrowIfCancellationRequested();
 
-                if (!library.IsValid(out IEnumerable<IError> errors))
+                ILibraryOperationResult result = await library.IsValidAsync(_dependencies);
```
Comment (jimmylewis) [inline src/LibraryManager/LibrariesValidator.cs]: ConfigureAwait

File: src/LibraryManager/Helpers/Extensions.cs [hunk 17]
```diff
@@ -19,45 +19,97 @@ namespace Microsoft.Web.LibraryManager
...
+                    else
+                    {
+                        ILibraryOperationResult result = await IsValidAsync(state, provider);
+                        errors.AddRange(result.Errors);
+                    }
+                }
+            }
+            else
+            {
+                errors.Add(PredefinedErrors.ManifestMalformed());
```
Comment (jimmylewis) [inline src/LibraryManager/Helpers/Extensions.cs]: >errors [](start = 16, length = 6)

Can we add an UnknownError?

File: src/LibraryManager/Helpers/Extensions.cs [hunk 18]
```diff
@@ -19,45 +19,97 @@ namespace Microsoft.Web.LibraryManager
...
+        /// <param name="dependencies">The <see cref="IDependencies"/> used to validate <see cref="ILibraryInstallationState"/></param>
+        /// <returns><see cref="ILibraryOperationResult"/> with the result of the validation</returns>
+        public static async Task<ILibraryOperationResult> IsValidAsync(this ILibraryInstallationState state, IDependencies dependencies)
         {
-            errors = null;
-            var list = new List<IError>();
+            var errors = new List<IError>();
 
-            if (state == null)
+            if (state != null)
```
Comment (jimmylewis) [inline src/LibraryManager/Helpers/Extensions.cs]: >if [](start = 12, length = 2)

Can we try to flatten this?

File: src/LibraryManager/Helpers/Extensions.cs [hunk 19]
```diff
@@ -19,45 +19,97 @@ namespace Microsoft.Web.LibraryManager
...
+                else if (state.DestinationPath.IndexOfAny(Path.GetInvalidPathChars()) >= 0)
+                {
+                    errors.Add(PredefinedErrors.DestinationPathHasInvalidCharacters(state.DestinationPath));
+                }
             }
-            else if (state.DestinationPath.IndexOfAny(Path.GetInvalidPathChars()) > 0)
+            else
             {
-                list.Add(PredefinedErrors.DestinationPathHasInvalidCharacters(state.DestinationPath));
+                errors.Add(PredefinedErrors.ManifestMalformed());
```
Comment (jimmylewis) [inline src/LibraryManager/Helpers/Extensions.cs]: >ManifestMalformed [](start = 44, length = 17)

UnknownError?

File: src/LibraryManager/Providers/Cdnjs/CdnjsProvider.cs [hunk 20]
```diff
@@ -237,35 +239,21 @@ public string GetSuggestedDestination(ILibrary library)
 
         private async Task<Stream> GetStreamAsync(ILibraryInstallationState state, string sourceFile, CancellationToken cancellationToken)
```
Comment (jimmylewis) [inline src/LibraryManager/Providers/Cdnjs/CdnjsProvider.cs]: >GetStreamAsync [](start = 35, length = 14)

We should make sure to be not on the UI thread here

File: src/LibraryManager/Providers/Cdnjs/CdnjsProvider.cs [hunk 21]
```diff
@@ -237,35 +239,21 @@ public string GetSuggestedDestination(ILibrary library)
...
 
-            string absolute = Path.Combine(CacheFolder, name, version, sourceFile);
-
-            if (File.Exists(absolute))
+            if (!string.IsNullOrEmpty(name) && !string.IsNullOrEmpty(version))
             {
-                return await HostInteraction.ReadFileAsync(absolute, cancellationToken).ConfigureAwait(false);
+                string absolute = Path.Combine(CacheFolder, name, version, sourceFile);
+
+                if (File.Exists(absolute))
```
Comment (jimmylewis) [inline src/LibraryManager/Providers/Cdnjs/CdnjsProvider.cs]: >File [](start = 20, length = 4)

Can you see if we hit this on the UI thread (e.g. with a UNC path)?

File: src/LibraryManager/Providers/Cdnjs/CdnjsProvider.cs [hunk 22]
```diff
@@ -319,35 +311,35 @@ public string GetLibraryVersion(string libraryId)
...
-
-            string cacheDir = Path.Combine(CacheFolder, name, version);
-            string destinationDir = Path.Combine(HostInteraction.WorkingDirectory, state.DestinationPath);
-
             try
             {
-                foreach (string sourceFile in state.Files)
+                (string name, string version) = LibraryNamingScheme.Instance.GetLibraryNameAndVersion(state.LibraryId);
+
+                if (!string.IsNullOrEmpty(name) && !string.IsNullOrEmpty(version))
```
Comment (jimmylewis) [inline src/LibraryManager/Providers/Cdnjs/CdnjsProvider.cs]: >if (!string.IsNullOrEmpty(name) && !string.IsNullOrEmpty(version)) [](start = 16, length = 66)

design follow up: Can we avoid doing this check in many places?

File: src/LibraryManager/Providers/Unpkg/UnpkgProvider.cs [hunk 23]
```diff
@@ -49,9 +50,10 @@ internal string CacheFolder
                 return LibraryOperationResult.FromCancelled(desiredState);
             }
 
-            if (!desiredState.IsValid(out IEnumerable<IError> errors))
+            ILibraryOperationResult validationResult = await desiredState.IsValidAsync(this);
```
Comment (jimmylewis) [inline src/LibraryManager/Providers/Unpkg/UnpkgProvider.cs]: >IsValidAsync [](start = 74, length = 12)

this should already be done in the manifest

File: src/LibraryManager/Providers/Cdnjs/CdnjsProvider.cs (hunk 22 above)
Comment (raquelsa) [reply src/LibraryManager/Providers/Cdnjs/CdnjsProvider.cs]: Logged issue 252

---
In reply to: [197970963](https://github.com/aspnet/LibraryManager/pull/230#discussion_r197970963) [](ancestors = 197970963)

File: src/LibraryManager/Providers/Cdnjs/CdnjsProvider.cs (hunk 20 above)
Comment (raquelsa) [reply src/LibraryManager/Providers/Cdnjs/CdnjsProvider.cs]: We are not on the main thread here.

---
In reply to: [197969429](https://github.com/aspnet/LibraryManager/pull/230#discussion_r197969429) [](ancestors = 197969429)

File: src/LibraryManager/Providers/Cdnjs/CdnjsProvider.cs (hunk 21 above)
Comment (raquelsa) [reply src/LibraryManager/Providers/Cdnjs/CdnjsProvider.cs]: The path is always a local path to the local cache folder

---
In reply to: [197970287](https://github.com/aspnet/LibraryManager/pull/230#discussion_r197970287) [](ancestors = 197970287)

File: src/LibraryManager/Manifest.cs [hunk 24]
```diff
@@ -452,56 +453,60 @@ public void AddVersion(string version)
...
-                                    files.Add(fileIdentifier);
-                                }
-                            }
-                        }
-                    }
+                    tasks.Add(GetFilesWithVersionsAsync(state));
                 }
+
+                IEnumerable<FileIdentifier>[] allFiles = await Task.WhenAll(tasks).ConfigureAwait(false);
+                files = allFiles.SelectMany(f => f).Distinct().ToList();
```
Comment (jimmylewis) [inline src/LibraryManager/Manifest.cs]: >.ToList() [](start = 62, length = 9)

nit: If files was declared as an IEnumerable above, you wouldn't need to call ToList() here.

File: src/LibraryManager/Manifest.cs [hunk 25]
```diff
@@ -452,56 +453,60 @@ public void AddVersion(string version)
...
-                    filesWithVersions = desiredStateFiles.Select(f => new FileIdentifier(Path.Combine(state.DestinationPath, f), library.Version));
+                    ILibraryOperationResult updatedStateResult = await provider.UpdateStateAsync(state, CancellationToken.None).ConfigureAwait(false);
+
+                    if (updatedStateResult.Success)
+                    {
+                        ILibrary library = await catalog.GetLibraryAsync(state.LibraryId, CancellationToken.None).ConfigureAwait(false);
+
+                        if (library != null && library.Files != null)
+                        {
+                            IEnumerable<string> desiredStateFiles = updatedStateResult.InstallationState.Files?.Where(f => library.Files.Keys.Contains(f));
```
[the remark this answers is not in the document]
Comment (prafullbhosale) [reply src/LibraryManager/Manifest.cs]: The `UpdateStateAsync` method is in the provider and the manifest should ideally not trust a provider to do validations. 
However, this check immediately fails for the `FilesystemProvider` which supports renaming files (i.e. the `GetLibraryAsync()` will return a library with the original file name, but the `UpdateStateAsync()` will return the target file name and they may not match.)

**Pull request merged**